homeassistant/components/husqvarna_automower/calendar.py

```diff
@@ -0,0 +1,114 @@
+"""Creates a calendar entity for the mower."""
+
+from datetime import datetime
+import logging
+
+from aioautomower.model import AutomowerCalendarEvent
+from ical.calendar import Calendar
+from ical.event import Event
+from ical.types.recur import Recur
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import AutomowerConfigEntry
+from .coordinator import AutomowerDataUpdateCoordinator
+from .entity import AutomowerBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: AutomowerConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up lawn mower platform."""
+    coordinator = entry.runtime_data
+    async_add_entities(
+        AutomowerCalendarEntity(mower_id, coordinator) for mower_id in coordinator.data
+    )
+
+
+class AutomowerCalendarEntity(AutomowerBaseEntity, CalendarEntity):
+    """Representation of the Automower Calendar element."""
+
+    _attr_name: str | None = None
+
+    def __init__(
+        self,
+        mower_id: str,
+        coordinator: AutomowerDataUpdateCoordinator,
+    ) -> None:
+        """Set up AutomowerCalendarEntity."""
+        super().__init__(mower_id, coordinator)
+        self._attr_unique_id = mower_id
+        self.calendar = Calendar()
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the current or next upcoming event."""
+        self._automower_to_ical_event(self.mower_attributes.calendar.events)
+        return _ical_to_calendar_event(self.calendar.events[0])
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range.
+
+        This is only called when opening the calendar in the UI.
+        """
+        self.calendar = await hass.async_add_executor_job(Calendar)
+        self._automower_to_ical_event(self.mower_attributes.calendar.events)
+        events = self.calendar.timeline_tz(start_date.tzinfo).overlapping(
+            start_date,
+            end_date,
+        )
+        return [_ical_to_calendar_event(event) for event in events]
+
+    def _automower_to_ical_event(
+        self, event_list: list[AutomowerCalendarEvent]
```

Comment:
This function uses both `event_list` and `self.mower_attributes.calendar.events`, so pick one.